src/mpid/ch4/netmod/ucx/ucx_recv.h

```diff
@@ -31,6 +50,32 @@ MPL_STATIC_INLINE_PREFIX void MPIDI_UCX_recv_cmpl_cb(void *request, ucs_status_t
             rreq->status.MPI_TAG = MPIDI_UCX_get_tag(info->sender_tag);
             MPIR_STATUS_SET_COUNT(rreq->status, count);
         }
+
+        switch (MPIR_CVAR_UCX_NONCONTIG_RECV) {
```

Comment:
Suggest save the choice in the request (as `type`) so `MPIDI_UCX_recv` can make more complicated choice. For example, the `iov` method probably need consider the datatype sparsity.

Reply:
What are the CVAR values in this case? Just a boolean where UCX handles noncontig types or it doesn't?

Reply:
Let me clarify. I suggest to take the CVAR as is in the issuing call, but save the choice into the request so the completion callback does not depend on the CVAR. This allows the issuing call the flexibility of taking additional factors such as the datatype sparsity in addition to the CVAR to make packing choice.

Reply:
I went with a boolean CVAR that decided whether MPICH or UCX is responsible for handling noncontig recvs. Default is for MPICH to handle so the user bug in #6708 will not trigger. MPICH will use an IOV recv when the datatype is dense and a pack buffer otherwise.